### PR TITLE
Fix: Prevents interpreting run options as yarn options

### PR DESCRIPTION
**Pull request by arcanis**
**Summary**

Running the following command doesn't have the expected effect, because the `--no-coverage` option is interpreter by Yarn instead of being forwarded to the script. This patch fix this.

```
$> yarn test-only --no-coverage     # currently bad
$> yarn test-only -- --no-coverage  # currently good
```

This is a breaking change: it will change some edge-case behaviors.

### CODE REVIEW

**Comment by BYK** [comment]
Also, tests or didn't happen.

**Comment by arcanis** [comment]
> Do you know if we have an open or closed issue around this?

No, this is an issue reported by the Oculus people, and something I personally struggle with on a daily basis. :p

> The problem with this approach is it makes yarn take no arguments at all such as --prefix or --modules-folder.

Not really an issue since the only potential option concerned is `--prefix`, which is not implemented and has its own set of issues + the solution is obvious: any option before the script name would have to be yarn options. We currently don't need this behavior, tho.

**Comment by kaylieEB** [comment]
Although it's convenient to not have to use the `--`, this would change the behavior from `npm` and how we write scripts inside `package.json`, so I'd prefer it if we could keep them the same.

**Comment by arcanis** [comment]
Hm I don't know, this behavior really is a pain, especially since it is completely silent, which means that unless people actually look at the command executed, they won't see that the options have been ignored. It's potentially dangerous (just imagine something like `yarn rm / --dry-run` being executed).

A less invasive possibility would be to only add this behavior when running `yarn <pkg>`. Running `yarn run <pkg>` would still require the `--`. I don't really like this because it causes an unnecessary split between the two commands behaviors, but maybe it's still better than the alternative.

**Comment by cpojer** [comment]
I think I'm on board with this change, but agree that we need tests for it. We had an internal discussion about this at FB previously and would like to make sure other people are involved in a vote for this feature. I seem to remember that there were some issues with this approach that made this less workable, but I do think that `yarn --options run --foo` makes a lot of sense to overall and will definitely be a better UX for beginners. Can we make this work at all in a backwards compatible way at all? 

@kaylieEB I think diverging here is ok, given that in scripts you'd write "yarn run …" instead of "npm run" or a generic command, like just "run", so it's clear that you are using either tool.

**Comment by arcanis** [comment]
I don't think we can - the only way I could see this working would be to add some kind of "if the command line doesn't begins with --, add one" (so that we wouldn't break people already using --), but that seems extremely confusing and unexpected.

**Comment by BYK** [comment]
I think the point we ended up is pretty great so 👍 . Thanks everyone, especially @arcanis. We need to be very loud and clear about this change though.

**Comment by cpojer** [comment]
I think that's confusing and will break projects with people on different versions of Yarn. Can we add a deprecation warning to 1.0 if the command starts with `<bin> run --` and continue on with the old behavior? I understand it is confusing, but I think breaking this in the 1.0 release will cause more pain than it provides value.

**Comment by kaylieEB** [comment]
> Can we add a deprecation warning to 1.0 if the command starts with <bin> run -- and continue on with the old behavior?

I think if this could be supported, it would be great since devs don't have to alter their current use habits, similar to how we support both `yarn [command]` and `yarn run [command]`. If not, definitely a big warning would be helpful.

**Comment by arcanis** [comment]
@BYK @kaylieEB good for review

File: src/reporters/lang/en.js [hunk 1]
```diff
@@ -188,6 +188,8 @@ const messages = {
 
   execMissingCommand: 'Missing command name.',
 
+  dashDashDeprecation:
+    "Using -- to pass arguments to your scripts isn't required anymore. Doing this may cause issues in future versions.",
```
Comment (cpojer) [inline src/reporters/lang/en.js]: How about: "Yarn does not require \`--\` to pass arguments to scripts. In a future version of Yarn, passing \`--\` will fail with an error." ?

File: src/cli/index.js [hunk 2]
```diff
@@ -403,7 +418,7 @@ export default function start() {
   const doubleDashIndex = process.argv.findIndex(element => element === '--');
   const startArgs = process.argv.slice(0, 2);
   const args = process.argv.slice(2, doubleDashIndex === -1 ? process.argv.length : doubleDashIndex);
-  const endArgs = doubleDashIndex === -1 ? [] : process.argv.slice(doubleDashIndex + 1, process.argv.length);
+  const endArgs = doubleDashIndex === -1 ? [] : process.argv.slice(doubleDashIndex, process.argv.length);
```
Comment (kaylieEB) [inline src/cli/index.js]: can you do `process.argv.slice(doubleDashIndex)` since it'll just default to length of the array?

File: src/reporters/lang/en.js (hunk 1 above)
Comment (arcanis) [inline src/reporters/lang/en.js]: It won't fail, it will just be forwarded to the binary as-is.

```
$> yarn test -- --no-coverage
```

will become

```
$> jest -- --no-coverage
```

instead of

```
$> jest --no-coverage
```

This is important because in some case you really want to prevent something from being interpreted as an option.

File: src/cli/index.js [hunk 3]
```diff
@@ -110,6 +110,16 @@ export function main({
     command = commands.run;
   }
 
+  let warnAboutRunDashDash = false;
+  // we using "yarn <script> -abc" or "yarn run <script> -abc", we want -abc to be script options, not yarn options
+  if (command === commands.run) {
+    if (endArgs.length === 0) {
+      endArgs = ['--', ...args.splice(1, args.length)];
```
Comment (kaylieEB) [inline src/cli/index.js]: same for here? `args.splice(1)`?

File: src/reporters/lang/en.js (hunk 1 above)
Comment (cpojer) [inline src/reporters/lang/en.js]: Right, in that case then, maybe this: "From Yarn 1.0, scripts do not require \`--\` to be passed. In a future version of Yarn, passing \`--\` will be forwarded to the script directly." ?

Comment (arcanis) [inline src/reporters/lang/en.js]: Much better than the current text, thanks 👍

File: src/cli/index.js (hunk 2 above)
Comment (arcanis) [inline src/cli/index.js]: Sure thing 👍

File: __tests__/integration.js [hunk 4]
```diff
@@ -73,6 +73,62 @@ test('--mutex network', async () => {
...
+
+    const command = path.resolve(__dirname, '../bin/yarn');
+    const options = {cwd, env: {YARN_SILENT: 1}};
+
+    const {stderr: stderr, stdout: stdout} = execa(command, ['run', 'echo', '--opt'], options);
+
+    const stdoutPromise = misc.consumeStream(stdout);
+    const stderrPromise = misc.consumeStream(stderr);
+
+    const [stdoutOutput, stderrOutput] = await Promise.all([stdoutPromise, stderrPromise]);
```
Comment (BYK) [inline __tests__/integration.js]: Is there an advantage to this indirection instead of just doing
```
const stdoutOutput = await misc.consumeStream(stdout);
const stderrOutput = await misc.consumeStream(stderr);
```

File: __tests__/integration.js [hunk 5]
```diff
@@ -73,6 +73,62 @@ test('--mutex network', async () => {
...
+
+    expect(stdoutOutput.toString().trim()).toEqual('--opt');
```
Comment (BYK) [inline __tests__/integration.js]: Are we testing the "$ <command>" echo string here? If so we may wanna add a comment since this may change.

File: __tests__/integration.js [hunk 6]
```diff
@@ -73,6 +73,62 @@ test('--mutex network', async () => {
...
+
+    const command = path.resolve(__dirname, '../bin/yarn');
+    const options = {cwd, env: {YARN_SILENT: 1}};
+
+    const {stderr: stderr, stdout: stdout} = execa(command, ['run', 'echo', '--', '--opt'], options);
+
+    const stdoutPromise = misc.consumeStream(stdout);
+    const stderrPromise = misc.consumeStream(stderr);
+
+    const [stdoutOutput, stderrOutput] = await Promise.all([stdoutPromise, stderrPromise]);
```
Comment (BYK) [inline __tests__/integration.js]: Lots of repeated code here. You may clean it up with `beforeEach` and `afterEach` hooks with a `describe('run with arguments', ` block.

File: src/cli/index.js [hunk 7]
```diff
@@ -110,6 +110,16 @@ export function main({
     command = commands.run;
   }
 
+  let warnAboutRunDashDash = false;
+  // we using "yarn <script> -abc" or "yarn run <script> -abc", we want -abc to be script options, not yarn options
```
Comment (BYK) [inline src/cli/index.js]: we _are_ using

File: src/cli/index.js [hunk 8]
```diff
@@ -171,6 +181,11 @@ export function main({
   //
   const run = (): Promise<void> => {
     invariant(command, 'missing command');
+
+    if (warnAboutRunDashDash) {
+      reporter.warn(reporter.lang('dashDashDeprecation'));
```
Comment (BYK) [inline src/cli/index.js]: Arguably, this should be an `info` message, not a `warning` since with the `--strict` option it will break and I don't think should break just because you are still using the harmless `--` thing. What do you think?

File: __tests__/integration.js (hunk 4 above)
Comment (arcanis) [inline __tests__/integration.js]: Problem with this writing is that it doesn't work, because the stdin and stdout streams are started and stopped as soon as the event loop gets a chance, ie. before the second line gets executed. Before of this, the second `consumeStream` call never receives the `end` event, and never returns. The synchronous version makes sure we attach both of our listeners before allowing the event loop to continue.

File: __tests__/integration.js (hunk 5 above)
Comment (arcanis) [inline __tests__/integration.js]: No, because we run with `YARN_SILENT: 1`, so Yarn doesn't print the `yarn run` header.

File: __tests__/integration.js (hunk 4 above)
Comment (BYK) [inline __tests__/integration.js]: Got it. I think this deserves a comment. Even better, find a way that doesn't require this workaround (not necessarily in this patch). Maybe use something else rather than the stream consumer? This may also help: https://nodejs.org/api/stream.html#stream_readable_resume

File: src/cli/index.js (hunk 8 above)
Comment (arcanis) [inline src/cli/index.js]: Makes sense, but the PR that will implement the `--strict` feature will need to revisit all the loglevels anyway.

**Comment by arcanis** [comment]
@BYK Long term plan is to switch to a different command line parser more suited to our use cases. I have it on my backlog, but time has been a scarce resource lately :p

File: src/cli/index.js (hunk 7 above)
Comment (arcanis) [inline src/cli/index.js]: me hulk

File: src/cli/index.js (hunk 8 above)
Comment (BYK) [inline src/cli/index.js]: Eh, fair point. I'd still get certain things "right" from the start but up to you, not gonna block.

**Comment by BYK** [comment]
>I have it on my backlog, but time has been a scarce resource lately :p

Shall we make it explicit by creating an issue?

File: __tests__/integration.js (hunk 5 above)
Comment (BYK) [inline __tests__/integration.js]: Oh nice. Then what are we testing here? I'm confused :D

**Comment by arcanis** [comment]
@BYK https://github.com/yarnpkg/yarn/issues/4225
